components/script/script_thread.rs

```diff
@@ -1072,6 +1074,54 @@ impl ScriptThread {
         }
     }
 
+    /// Process a compositor mouse leave event.
+    fn process_mouse_leave_event(
+        &self,
+        document: &Document,
+        position: Point2D<f32, DevicePixel>,
+        pressed_mouse_buttons: u16,
+        can_gc: CanGc,
+    ) {
+        if let Some(target) = self.topmost_mouse_over_target.take() {
+            let window = document.window();
```

Comment:
`script_thread.rs` is already very big and everything from here is document specific, so please move it a new method called `Document::handle_mouse_leave_event`.